src/webgpu/api/validation/image_copy/layout_related.spec.ts

```diff
@@ -250,68 +254,86 @@ g.test('offset_alignment')
   });
 
 g.test('bound_on_bytes_per_row')
+  .desc(`For all formats, verify image copy validations w.r.t bytesPerRow.`)
   .params(u =>
     u
       .combine('method', kImageCopyTypes)
       .combine('format', kSizedTextureFormats)
       .filter(formatCopyableWithMethod)
       .beginSubcases()
-      .combineWithParams([
-        { blocksPerRow: 2, additionalPaddingPerRow: 0, copyWidthInBlocks: 2 }, // success
-        { blocksPerRow: 2, additionalPaddingPerRow: 5, copyWidthInBlocks: 3 }, // success if bytesPerBlock <= 5
-        { blocksPerRow: 1, additionalPaddingPerRow: 0, copyWidthInBlocks: 2 }, // failure, bytesPerRow > 0
-        { blocksPerRow: 0, additionalPaddingPerRow: 0, copyWidthInBlocks: 1 }, // failure, bytesPerRow = 0
-      ])
-      .combineWithParams([
-        { copyHeightInBlocks: 0, copyDepth: 1 }, // we don't have to check the bound
-        { copyHeightInBlocks: 1, copyDepth: 0 }, // we don't have to check the bound
-        { copyHeightInBlocks: 2, copyDepth: 1 }, // we have to check the bound
-        { copyHeightInBlocks: 0, copyDepth: 2 }, // we have to check the bound
-      ])
+      .expand('params', p => {
+        const info = kTextureFormatInfo[p.format];
+        return [
+          // Copying exact fit with aligned bytesPerRow should work.
+          {
+            bytesPerRow: 256,
+            blocksPerRow: 256 / info.bytesPerBlock,
+            copyWidthInBlocks: 256 / info.bytesPerBlock,
+            success: true,
+          },
+          // Copying larger texture when padding in bytesPerRow is enough should work.
+          {
+            bytesPerRow: 256,
+            blocksPerRow: 256 / info.bytesPerBlock - 1,
+            copyWidthInBlocks: 256 / info.bytesPerBlock,
+            success: true,
+          },
+          // Unaligned bytesPerRow should not work unless the method is 'WriteTexture'.
+          {
+            bytesPerRow: 257,
+            blocksPerRow: 256 / info.bytesPerBlock,
+            copyWidthInBlocks: 256 / info.bytesPerBlock,
+            success: p.method === 'WriteTexture',
+          },
+          // Copying larger texture when padding in bytesPerRow is not enough should fail.
+          {
+            bytesPerRow: 256,
+            blocksPerRow: 256 / info.bytesPerBlock,
+            copyWidthInBlocks: 256 / info.bytesPerBlock + 1,
+            success: false,
+          },
+          // When bytesPerRow is smaller than bytesInLastRow copying should fail.
```

Comment:
Don't this and the previous test hit the same validation rule? (That's fine, but the comments make it seem otherwise)

Reply:
I think they should be different? The first one checks that it fails because essentially out of bounds, and the second one should fail because bytesInLastRow = widthInBlocks * bytesPerBlock (512) > bytesPerRow (256).